blockchain/utxocache.go

```diff
@@ -725,22 +726,34 @@ func (b *BlockChain) InitConsistentState(tip *blockNode, interrupt <-chan struct
 // Example: if the last flush hash was at height 100 and one of the deleted blocks was at
 // height 98, this function will return true.
 func (b *BlockChain) flushNeededAfterPrune(deletedBlockHashes []chainhash.Hash) (bool, error) {
-	lastFlushHeight, err := b.BlockHeightByHash(&b.utxoCache.lastFlushHash)
-	if err != nil {
-		return false, err
-	}
+	node := b.index.LookupNode(&b.utxoCache.lastFlushHash)
+	if node == nil {
+		// If we couldn't find the node where we last flushed at, have the utxo cache
+		// flush to be safe and that will set the last flush hash again.
+		//
+		// This realistically should never happen as nodes are never deleted from
+		// the block index.  This happening likely means that there's a hardware
+		// error which is something we can't recover from.  The best that we can
+		// do here is to just force a flush and hope that the newly set
+		// lastFlushHash doesn't error.
+		return true, nil
+	}
+	lastFlushHeight := node.Height()
 
 	// Loop through all the block hashes and find out what the highest block height
 	// among the deleted hashes is.
 	highestDeletedHeight := int32(-1)
 	for _, deletedBlockHash := range deletedBlockHashes {
-		height, err := b.BlockHeightByHash(&deletedBlockHash)
-		if err != nil {
-			return false, err
+		node := b.index.LookupNode(&deletedBlockHash)
+		if node == nil {
+			// If we couldn't find this node, just skip it and try the next
```

Comment:
This comment doesn't seem consistent with the commit message, as a stale block might well happen, and this case is not necessarily a db corruption.

Reply:
So this is reading from `blockIndex`, which will store a mapping of `hash => height` for every block we've ever seen:
* https://github.com/btcsuite/btcd/blob/9ea2eea99bcb799d43f4164ebae41a61df90d133/blockchain/accept.go#L70-L74
* https://github.com/btcsuite/btcd/blob/9ea2eea99bcb799d43f4164ebae41a61df90d133/blockchain/chainio.go#L1382-L1402


We never delete from this index, so even if it's a stale block, if we attempt to accept it in the past, it _should_ be in this index. 

The issue with this code was that it used `BlockHeightByHash`, which only checks for block currently in the main chain: https://github.com/btcsuite/btcd/blob/9ea2eea99bcb799d43f4164ebae41a61df90d133/blockchain/chain.go#L1519

Before this would error out, and cause the entire daemon to be torn down.

Reply:
Right as roasbeef mentioned, since we never delete entries from the blockIndex, if we have the block on disk, the index should also have it.

Does this clarify the comment?